Bitbucket.Authentication/OAuth/OAuthAuthenticator.cs

```diff
@@ -0,0 +1,321 @@
+            ConsumerKey,
+            scope.ToString(),
+            CallbackUri);
+
+            //Open the browser for auth
+            var url = new Uri(new Uri("https://bitbucket.org"), authorityUrl).AbsoluteUri;
+            Process.Start(url);
+
+            string rawUrlData;
+            try
+            {
+                //Fetch the url and await for the reply
+                rawUrlData = await SimpleServer.WaitForURLAsync(CallbackUri, cancellationToken);
+            }
+            catch (Exception ex)
+            {
+                string message;
+                string details;
+                if (ex.InnerException != null && ex.InnerException.GetType().IsAssignableFrom(typeof(TimeoutException)))
+                {
+                    message = "Timeout awaiting response from Host service.";
+                    details = "Please check the SourceTree's configuration for this Host." + Environment.NewLine +
+                              "Confirm that if there are overrides to the OAuth Consumer Information that they are correct";
+                }
+                else
+                {
+                    message = "Unable to receive callback from OAuth service provider";
+                    details = "Please try restarting SourceTree";
+                }
+
+                //throw new OAuthAuthenticationFlowException(message, details, ex);
+                throw new Exception(message + details, ex);
+            }
+
+
+            //Parse and try to get the key
+            Dictionary<string,string> qs =
+                rawUrlData.Split('&')
+               .ToDictionary(c => c.Split('=')[0],
+                             c => Uri.UnescapeDataString(c.Split('=')[1]));
+            var authCode = qs.Keys.Where(k => k.EndsWith("code", StringComparison.InvariantCultureIgnoreCase)).Select(k => qs[k]).FirstOrDefault();
+
+
+            if (string.IsNullOrWhiteSpace(authCode))
+            {
+                var error_desc = qs["error_description"];
+                throw new Exception("Request for an OAuth request_token was denied" + error_desc);
+            }
+
+            return authCode;
+        }
+
+        /// <summary>
+        ///     Use a request_token to get an access_token
+        /// </summary>
+        /// <param name="targetUri"></param>
+        /// <param name="authCode"></param>
+        /// <returns></returns>
+        private async Task<AuthenticationResult> GetAccessToken(TargetUri targetUri, string authCode)
+        {
+            Token token = null;
+            Token refreshToken = null;
+
+            using (HttpClientHandler handler = targetUri.HttpClientHandler)
+            {
+                using (HttpClient httpClient = new HttpClient(handler)
+                {
+                    Timeout = TimeSpan.FromMilliseconds(RequestTimeout)
+                })
+                {
+                    var tokenUrl = string.Format(
```

Comment:
This is a fairly complex call to `String.Format`, I'd rather see string interpolation here.

Example:
```
$"{TokenUri}?grant_type=authorization_code&code={authCode}&client_id={ConsumerKey}&client_secret={ConsumerSecret}&state=authenticated"
```